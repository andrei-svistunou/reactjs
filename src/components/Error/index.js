import React from 'react';
import PropTypes from 'prop-types';

class Error extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      errorMSg: null
    };
  }

  componentDidCatch(error, errorMSg) {
    this.setState({
      error,
      errorMSg
    });
  }

  render() {
    if (this.state.error) {
      return (
        <div>
          <h2>{'Error...'}</h2>
          <p>
            {this.state.error && this.state.error.toString()}
          </p>
          <div>{'Component Stack Error Details: '}</div>
          <p>{this.state.errorMSg.componentStack}</p>
        </div>
      );
    }

    return this.props.children;
  }
}

Error.propTypes = {
  children: PropTypes.element,
};

export default Error;
