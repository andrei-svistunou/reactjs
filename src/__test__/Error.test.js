import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Error from '../components/Error';

beforeEach(() => {
  jest.spyOn(console, 'error');
  global.console.error.mockImplementation(() => {});
});

afterEach(() => {
  global.console.error.mockRestore();
});

function ProblemChild() {
  throw new Error('Error thrown from problem child');
  return <div>Error</div>; // eslint-disable-line
}

describe('<Error />', () => {
  it('should catch errors with componentDidCatch', () => {
    const spy = jest.spyOn(Error.prototype, 'componentDidCatch');
    const err = mount(<Error><ProblemChild /></Error>);
    expect(Error.prototype.componentDidCatch).toHaveBeenCalled();
    expect(toJson(err)).toMatchSnapshot();
  });
});
