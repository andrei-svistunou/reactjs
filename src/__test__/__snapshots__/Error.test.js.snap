// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Error /> should catch errors with componentDidCatch 1`] = `
<Error>
  <div>
    <h2>
      Error...
    </h2>
    <p>
      [object Object]
    </p>
    <div>
      Component Stack Error Details: 
    </div>
    <p>
      
    in ProblemChild
    in Error (created by WrapperComponent)
    in WrapperComponent
    </p>
  </div>
</Error>
`;
